// @shopify/react-async/dist/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var async_1 = require("@shopify/async");
exports.DeferTiming = async_1.DeferTiming;
tslib_1.__exportStar(require("./types"), exports);
tslib_1.__exportStar(require("./hooks"), exports);
var Prefetcher_1 = require("./Prefetcher");
exports.Prefetcher = Prefetcher_1.Prefetcher;
var PrefetchRoute_1 = require("./PrefetchRoute");
exports.PrefetchRoute = PrefetchRoute_1.PrefetchRoute;
var component_1 = require("./component");
exports.createAsyncComponent = component_1.createAsyncComponent;
var provider_1 = require("./provider");
exports.createAsyncContext = provider_1.createAsyncContext;
var assets_1 = require("./context/assets");
exports.AsyncAssetContext = assets_1.AsyncAssetContext;
exports.AsyncAssetManager = assets_1.AsyncAssetManager;
var prefetch_1 = require("./context/prefetch");
exports.PrefetchContext = prefetch_1.PrefetchContext;
exports.PrefetchManager = prefetch_1.PrefetchManager;
